From 9d99096193716d5dad7cac3dd982c3f66ed0a786 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Mon, 12 Jan 2026 17:24:37 +0100
Subject: [PATCH 950/954] fix(st): rework metadata partition selection
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The current implementation relies only on the partition type GUID, which
is not sufficient to select the correct metadata partition. Use both the
partition type GUID and the partition GUID instead, and improve warning
messages accordingly.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 41575b25c972b6f8958c101458d06ed23e38442b fix(st): rework metadata partition selection]
Change-Id: I1ca52b04fe85c7f5b0e1ce6c9c7658b68e692b98
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/523845
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 plat/st/common/bl2_io_storage.c | 55 ++++++++++++++++++++++++---------
 1 file changed, 41 insertions(+), 14 deletions(-)

diff --git a/plat/st/common/bl2_io_storage.c b/plat/st/common/bl2_io_storage.c
index 9a02edfe9b..600bd84af3 100644
--- a/plat/st/common/bl2_io_storage.c
+++ b/plat/st/common/bl2_io_storage.c
@@ -939,6 +939,43 @@ void plat_fwu_set_images_source(const struct fwu_metadata *metadata)
 	}
 }
 
+#if (STM32MP_SDMMC || STM32MP_EMMC)
+static int get_metadata_partition_entry(unsigned int image_id,
+					const partition_entry_t **entry)
+{
+	const bool is_main_metadata = (image_id == FWU_METADATA_IMAGE_ID);
+	const struct efi_guid metadata_type_guid = FWU_METADATA_GUID;
+	const struct efi_guid fwu_metadata1_part_guid = FWU_METADATA1_PART_GUID;
+	const struct efi_guid fwu_metadata2_part_guid = FWU_METADATA2_PART_GUID;
+	const struct efi_guid *metadata_part_guid =
+		is_main_metadata ?
+		&fwu_metadata1_part_guid : &fwu_metadata2_part_guid;
+
+	const char *metadata_part_name =
+		is_main_metadata ? METADATA_PART_1 : METADATA_PART_2;
+
+	partition_init(GPT_IMAGE_ID);
+
+	*entry = get_partition_entry_by_type_and_guid(&metadata_type_guid,
+						      metadata_part_guid);
+	if (*entry != NULL) {
+		return 0;
+	}
+
+	VERBOSE("FWU %s partition GUID not found, trying lookup by partition name...\n",
+		metadata_part_name);
+
+	*entry = get_partition_entry(metadata_part_name);
+	if (*entry == NULL) {
+		ERROR("Unable to find metadata partition \"%s\"\n",
+		      metadata_part_name);
+		return -ENOENT;
+	}
+
+	return 0;
+}
+#endif
+
 static int plat_set_image_source(unsigned int image_id,
 				 uintptr_t *handle,
 				 uintptr_t *image_spec)
@@ -947,7 +984,7 @@ static int plat_set_image_source(unsigned int image_id,
 	io_block_spec_t *spec __maybe_unused;
 	const partition_entry_t *entry __maybe_unused;
 	const uint16_t boot_itf = stm32mp_get_boot_itf_selected();
-	const struct efi_guid metadata_type_guid __maybe_unused = FWU_METADATA_GUID;
+	int ret __maybe_unused;
 
 	policy = &policies[image_id];
 	spec = (io_block_spec_t *)policy->image_spec;
@@ -956,19 +993,9 @@ static int plat_set_image_source(unsigned int image_id,
 #if (STM32MP_SDMMC || STM32MP_EMMC)
 	case BOOT_API_CTX_BOOT_INTERFACE_SEL_FLASH_SD:
 	case BOOT_API_CTX_BOOT_INTERFACE_SEL_FLASH_EMMC:
-		partition_init(GPT_IMAGE_ID);
-
-		entry = get_partition_entry_by_type(&metadata_type_guid);
-		if (entry == NULL) {
-			entry = (image_id == FWU_METADATA_IMAGE_ID) ?
-				get_partition_entry(METADATA_PART_1) :
-				get_partition_entry(METADATA_PART_2);
-
-
-			if (entry == NULL) {
-				ERROR("Unable to find a metadata partition\n");
-				return -ENOENT;
-			}
+		ret = get_metadata_partition_entry(image_id, &entry);
+		if (ret != 0) {
+			return ret;
 		}
 
 		spec->offset = entry->start;
-- 
2.53.0

